File: src/agents/claude-adapter.ts
/**
 * Claude adapter — implements Provider interface using @anthropic-ai/sdk.
 * Streams responses via the messages API.
 */

import Anthropic from "@anthropic-ai/sdk";
import type { Provider } from "./provider.js";
import type { Message, StreamOptions } from "../types.js";

const DEFAULT_MODEL = "claude-sonnet-4-20250514";

export class ClaudeAdapter implements Provider {
  private client: Anthropic;

  constructor(apiKey?: string) {
    this.client = new Anthropic({
      apiKey: apiKey || process.env.ANTHROPIC_API_KEY,
    });
  }

  async *stream(
    messages: Message[],
    options: StreamOptions
  ): AsyncIterable<string> {
    const anthropicMessages = messages.map((m) => ({
      role: m.role as "user" | "assistant",
      content: m.content,
    }));

    const stream = this.client.messages.stream({
      model: options.model || DEFAULT_MODEL,
      max_tokens: options.maxTokens,
      system: options.systemPrompt,
      messages: anthropicMessages,
    });

    for await (const event of stream) {
      if (
        event.type === "content_block_delta" &&
        event.delta.type === "text_delta"
      ) {
        yield event.delta.text;
      }
    }
  }
}

export function validateApiKey(): boolean {
  return !!process.env.ANTHROPIC_API_KEY;
}

File: redteam_arena/agents/claude_adapter.py
"""
Claude adapter -- implements Provider interface using the anthropic Python SDK.
Streams responses via the messages API.
"""

from __future__ import annotations

import os
from collections.abc import AsyncIterator

import anthropic

from redteam_arena.agents.provider import Provider
from redteam_arena.types import Message, StreamOptions

DEFAULT_MODEL = "claude-sonnet-4-20250514"


class ClaudeAdapter(Provider):
    def __init__(self, api_key: str | None = None, model: str = "") -> None:
        self._model = model
        self._client = anthropic.Anthropic(
            api_key=api_key or os.environ.get("ANTHROPIC_API_KEY", ""),
        )

    async def stream(
        self, messages: list[Message], options: StreamOptions
    ) -> AsyncIterator[str]:
        anthropic_messages = [
            {"role": m.role, "content": m.content} for m in messages
        ]

        with self._client.messages.stream(
            model=options.model or self._model or DEFAULT_MODEL,
            max_tokens=options.max_tokens,
            system=options.system_prompt,
            messages=anthropic_messages,  # type: ignore[arg-type]
        ) as stream:
            for text in stream.text_stream:
                yield text


def validate_api_key() -> bool:
    return bool(os.environ.get("ANTHROPIC_API_KEY"))
